### PR TITLE
feat(eval): add evaluation harness with ship criteria and regression tracking

### PR DESCRIPTION
**Pull request by William-Hill**
## Summary

Sprint 3 of the fine-tuned model epic (#124). Adds an automated evaluation harness that runs held-out test sets through fine-tuned models, computes per-adapter metrics, checks ship/no-ship criteria, and stores results for regression tracking.

- **Ship criteria** — codifies blocking/non-blocking thresholds from design spec Section 6.4 for all 3 adapters (parser, explainer, evaluator), with `--partial` mode for development
- **Metric computation** — entity F1, JSON/schema validity rates, latency percentiles, hallucination accuracy, relevance MAE per adapter
- **`ModelEvalRun` database model** — persists eval runs with metrics (JSONB), ship decision, and test set hash for version-over-version tracking
- **Eval harness CLI** — `python -m scripts.training.run_eval_harness` loads test JSONL, runs models, computes metrics, checks criteria, optionally persists to DB
- **Regression detection** — compares current vs previous metrics with tolerance threshold, direction-aware alerts (latency increase vs accuracy decrease)

## Test plan

- [x] 220 training tests passing, 572 total tests passing, 0 failures
- [x] Test sets load correctly (15 parser, 20 explainer, 15 evaluator examples)
- [x] Ship criteria match design spec Section 6.4 exactly
- [ ] Run `python -m scripts.training.run_eval_harness --task query_parser` with Ollama models loaded
- [ ] Run with `--persist` flag to verify DB writes

### Known limitations
- Explainer/evaluator report `no_ship` without `--partial` flag (LLM-judged metrics deferred)
- Test set sizes smaller than spec targets (15-20 vs 30-50)
- `adversarial_pass_rate`, `community_framing_f1`, `relevance_correlation` deferred

## Summary by CodeRabbit

* **New Features**
  * Added an automated model evaluation CLI that runs task-specific evaluations (parser, explainer, evaluator), computes quality metrics (entity matching, JSON-valid rates, latency p50/p95, hallucination accuracy, relevance error), detects regressions, and emits ship/ no-ship decisions with a human-readable multi-section report highlighting blocking vs non-blocking failures.

* **Chores**
  * Added persistent storage and indexes for storing evaluation runs and results.

* **Tests**
  * Added unit tests covering metrics, reporting, regression detection, and persistence.

### CODE REVIEW

File: scripts/training/run_eval_harness.py [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+            continue
+        prev_val = previous.get(metric)
+        if prev_val is None:
+            continue
+
+        delta = cur_val - prev_val
+
+        if metric in _HIGHER_IS_WORSE:
+            # For latency/MAE, increase is bad
+            if delta > tolerance:
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply scripts/training/run_eval_harness.py]: Won't fix — the tolerance is intentionally absolute. For 0-1 rate metrics (json_valid_rate, entity_f1), 0.02 = 2 percentage points which is a reasonable significance threshold. For latency (500-3000ms range), +0.02ms is below noise. The current behavior is correct for expected metric ranges.

File: scripts/training/run_eval_harness.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+        expected_labels = []
+        for ex in test_set:
+            exp = ex.get("expected") or {}
+            label = exp.get("label", "")
+            expected_labels.append(label)
+        metrics = compute_evaluator_metrics(
+            outputs, expected_labels=expected_labels if any(expected_labels) else None
+        )
+    else:
+        metrics = {}
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply scripts/training/run_eval_harness.py]: Fixed in c8e212c9. Added `logger.warning("Unknown task '%s', returning empty metrics", task)`.

File: scripts/training/run_eval_harness.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+        help="Ollama base URL",
+    )
+    parser.add_argument(
+        "--persist", action="store_true",
+        help="Save results to model_eval_runs DB table",
+    )
+    parser.add_argument(
+        "--partial", action="store_true",
+        help="Only check computable metrics (skip LLM-judged criteria)",
+    )
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply scripts/training/run_eval_harness.py]: Fixed in c8e212c9. `args.partial` is now passed through `run_task_eval(partial=args.partial)` → `check_ship_criteria(checkable, task, partial=partial)`. In partial mode, missing LLM-judged metrics are skipped instead of treated as failures.

File: supabase/migrations/20260325000001_add_model_eval_runs.sql [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    test_set_hash VARCHAR(64) NOT NULL,
+    metrics JSONB NOT NULL,
+    ship_decision VARCHAR(20) NOT NULL,
+    blocking_failures JSONB,
+    created_at TIMESTAMPTZ NOT NULL DEFAULT now()
+);
+
+CREATE INDEX IF NOT EXISTS idx_model_eval_runs_model_name ON model_eval_runs (model_name);
+CREATE INDEX IF NOT EXISTS idx_model_eval_runs_task ON model_eval_runs (task);
+CREATE INDEX IF NOT EXISTS idx_model_eval_runs_created_at ON model_eval_runs (created_at);
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply supabase/migrations/20260325000001_add_model_eval_runs.sql]: Fixed in c8e212c9. Added composite index `idx_model_eval_runs_model_task_created_at ON model_eval_runs (model_name, task, created_at DESC)`.

File: scripts/training/eval_harness.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        try:
+            expected = json.loads(assistant_msg)
+        except (json.JSONDecodeError, KeyError):
+            expected = None
+        results.append({
+            "input": user_msg,
+            "expected_raw": assistant_msg,
+            "expected": expected,
+        })
+    return results
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply scripts/training/eval_harness.py]: Won't fix — the test sets are generated by our own pipeline and always have the 3-message ChatML format. Adding defensive checks for malformed input here adds complexity for a scenario that indicates a broken data pipeline, not a harness bug. A JSONDecodeError or IndexError from a corrupted test file is appropriate failure behavior.

File: scripts/training/eval_harness.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    return {
+        "hallucination_accuracy": (
+            correct_labels / label_count if label_count > 0 else 0.0
+        ),
+        "relevance_mae": (
+            total_score_error / scored_count if scored_count > 0 else None
+        ),
+        "bias_mae": None,  # Requires separate bias test set
+        "relevance_correlation": None,  # Requires scipy — deferred
+    }
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply scripts/training/eval_harness.py]: By design. The hallucination label comparison uses `.get("label", "").upper()` — if the model omits the label field, it defaults to empty string which won't match any expected label, correctly counting as an incorrect prediction. No false positives possible.
